add inference lattice element for Tuples

jrevels opened the pull request:
w/ @vtjnash 

There are probably still some cases to handle and tests to add here...this is just a dump of what we currently have.

Once this PR is complete, it should allow for more robust constant prop of vargs/tuple elements.

jrevels commented:
here's a case still needs to updated: https://github.com/JuliaLang/julia/pull/28955/files#diff-493d285f483504268eb7cf24ea13b9aaR174

jrevels commented on base/compiler/inferenceresult.jl:
```diff
@@ -13,32 +13,27 @@ mutable struct InferenceResult
             result = linfo.rettype
         end
-        return new(linfo, EMPTY_VECTOR, result, nothing)
+        return new(linfo, compute_inf_result_argtypes(linfo), result, nothing)
```
Originally, I was going to basically rename `get_argtypes(::InferenceResult)` to `init_argtypes`,  but then I couldn't find a place where it ever looked like we wanted to construct an InferenceResult without initializing the `argtypes`. Hence this change, and getting rid of `get_argtypes(::InferenceResult)` altogether. 

Am I correct here that - currently, anyway - all cases where we construct an `InferenceResult`, we also want to initialize `argtypes`?

JeffBezanson commented:
I'm looking at this. @jrevels can you provide a test case or two?

JeffBezanson commented on base/compiler/ssair/inlining.jl:
```diff
@@ -808,7 +803,9 @@ function assemble_inline_todo!(ir::IRCode, linetable::Vector{LineInfoNode}, sv::
             for i = 3:length(atypes)
-                typ = widenconst(atypes[i])
+                typ = atypes[i]
+                typ isa PartialTuple && continue
```
This looks wrong; a `PartialTuple` can still be vararg I think?

JeffBezanson commented on base/compiler/inferenceresult.jl:
```diff
@@ -4,90 +4,89 @@ const EMPTY_VECTOR = Vector{Any}()
-    vargs = nothing
+    if !toplevel && caller_argtypes !== nothing
+        for i in 1:length(caller_argtypes)
+            a = maybe_widen_conditional(given_argtypes[i])
```
Is this meant to be `caller_argtypes[i]`?

jrevels replied:
Ah, yup! Thanks for that catch.

jrevels commented:
I *think* this test case is at least representative of the issue this PR seeks to resolve (e.g. https://github.com/jrevels/Cassette.jl/issues/71), though it might end up being e.g. necessary but insufficient:

```julia
julia> a(f, args...) = f(args...)
a (generic function with 1 method)

julia> b(args::Tuple) = a(args...)
b (generic function with 1 method)

julia> c(args...) = b(args)
c (generic function with 1 method)

julia> d(f, x, y) = c(f, Bool, x, y)
d (generic function with 1 method)

julia> f(::Type{Bool}, x, y) = x
f (generic function with 1 method)

julia> f(::DataType, x, y) = y
f (generic function with 2 methods)

# I want this to infer exactly Int64 as the output type
julia> @code_typed optimize=false d(f, 1, 2.0)
CodeInfo(
1 1 ─ %1 = (Main.c)(f, Main.Bool, x, y)::Union{Float64, Int64}                                                    │
  └──      return %1                                                                                              │
) => Union{Float64, Int64}
```

jrevels commented:
Okay, I added the test case I constructed above, which now passes. Made some other changes based on @JeffBezanson's feedback (thanks!) that lets a bunch of other failing cases pass now. Let's see if CI passes. 

In the meantime, I'll see if this fixes the Cassette problem cases that motivated this...

jrevels commented:
Okay, looks like this PR does resolve the OP in https://github.com/jrevels/Cassette.jl/issues/71, but even with this fix Cassette still requires [those hacky manual primitives just to force specialization](https://github.com/jrevels/Cassette.jl/issues/71#issuecomment-412093864) in order to pass [its inferrability tests](https://github.com/jrevels/Cassette.jl/blob/master/test/misctests.jl#L319). 

The overarching goal is to get rid of those entirely, but at least this is an improvement.

jrevels commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

jrevels commented:
:( no nanosoldier?

jrevels commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

...let's try again

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/28955#issuecomment-423269495) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/0ed9fbfb62550a79baa169693c83ab3120868c81/bc3489d_vs_6354405/report.md). cc @ararslan

jrevels commented:
Oof. Was definitely hoping to see some better perf improvements here. Instead, it looks like this causes a bunch of regressions...I'll do some debugging tomorrow.

jrevels commented:
So @vtjnash helped me fix the performance problems from the previous benchmarks - it turns out I was using `Core.Typeof` where I was supposed to use `typeof`, which coincidentally caused my test case here to pass but broke a bunch of other stuff. Fixing that fixed the benchmarks we tried, but now my original test case no longer passes.

The [most recent commit](https://github.com/JuliaLang/julia/pull/28955/commits/437334bbda13b328497aa18ff93b1a4c7b9498ac) is an attempt at more correctly handling the better type info we have floating around the cache. I skipped CI on it, though, because I messed something up somewhere and am getting a bootstrap `BoundsError` during compaction:

```
LoadError(at "compiler/compiler.jl" line 3: LoadError(at "compiler/bootstrap.jl" line 8: BoundsError(a=(), i=1)))
⋮
isassigned at ./array.jl:205
iterate at ./compiler/ssair/ir.jl:1118
iterate at ./compiler/ssair/ir.jl:1054
foreach at ./abstractarray.jl:1844
compact! at ./compiler/ssair/ir.jl:1254
run_passes at ./compiler/ssair/driver.jl:117
optimize at ./compiler/optimize.jl:160
typeinf at ./compiler/typeinfer.jl:35
```

jrevels commented on base/compiler/inferenceresult.jl:
```diff
@@ -4,90 +4,107 @@ const EMPTY_VECTOR = Vector{Any}()
+        isva_result_argtypes[nargs] = tuple_tfunc(result_argtypes[(nargs - 1):end])
+        return isva_result_argtypes
+    end
+    return result_argtypes
```
hmmm replacing this whole method body with a call to matching_cache_argtypes(linfo, nothing) at least allows it to build past the point that I was getting a BoundsError before (though of course probably results in a bunch of cache misses)

jrevels commented:
Okay, had a stupid off-by-one error. It builds AND it passes the above test now. 

However, this PR no longer resolves https://github.com/jrevels/Cassette.jl/issues/71. 

Here's another MWE-ish test for this PR then:

```julia
julia> function g(x, y)
           _1 = Core.tuple(Main.Bool)
           _2 = isa(y, Int) ? nothing : _1
           return tuple(_1..., x...)
       end
g (generic function with 1 method)

julia> function h(x, ::Type{S}) where S
           T = getfield(g(x, 1.0), 1)
           if T <: S
               return 1
           else
               return 1.0
           end
       end
h (generic function with 1 method)

# should infer as Int64
julia> @code_typed h((1,), Bool)
CodeInfo(
2 1 ─ %1 = Main.Bool::Type{Bool}                                                                               │╻ g
3 │   %2 = (%1 <: $(Expr(:static_parameter, 1)))::Bool                                                         │
  └──      goto #3 if not %2                                                                                   │
4 2 ─      return 1                                                                                            │
6 3 ─      return 1.0                                                                                          │
) => Union{Float64, Int64}
```

Anyway, let's see how CI goes. If it passes and we run benchmarks again and things are cool, it might be better to merge this and then do a follow-up PR.

Keno commented:
Trying out this PR, I found the following useful:
```
diff --git a/base/compiler/tfuncs.jl b/base/compiler/tfuncs.jl
index b15a2cc1f8..e2b3156bd1 100644
--- a/base/compiler/tfuncs.jl
+++ b/base/compiler/tfuncs.jl
@@ -416,6 +416,13 @@ function typeof_tfunc(@nospecialize(t))
         else
             return Type
         end
+    elseif isa(t, PartialTuple)
+        t = Tuple{anymap(widenconst, t.fields)...}
+        if isconcretetype(t) || isvarargtype(t)
+            return Const(t)
+        else
+            return Type{<:t}
+        end
     elseif isa(t, DataType)
         if isconcretetype(t) || isvarargtype(t)
             return Const(t)
```
I also took a quick look at the assertion failure and it seems to be doing the equivalent of `trunc(Int32, x::Int8)` which asserts in LLVM. We probably need to emit a check in codegen and bail out in that case.

vtjnash commented:
The `typeof(t) <: t.typ`, so the tfunc for that actually is:
```julia
t = t.typ # aka `t = widenconst(t)`
isconcretetype(t) && return Const(t)
return Type{<:t}
```

(all that business with `widenconst` will just give back wrong answers, we already cache the correct answer in `t.typ` since it's needed pretty frequently)

jrevels commented:
So @vtjnash and I spent a few hours today trying to debug the assertion failure here. It was a struggle, and we didn't totally succeed in figuring it out yet, but here's what we have so far.

You can trigger the assertion failure simply by building with `LLVM_ASSERTIONS = 1`; the miscompile triggering the assertion failure is from this code: https://github.com/JuliaLang/julia/blob/f5cdfd864d8c38e4134d1b8edbd587d12fe5b1de/base/process.jl#L348 

Matching what Keno already found, this resulting IR from compiling this code contains a nonsense call to `trunc(Int32, 0x02)`. Jameson and I guessed that this was caused by inlining a bad cache lookup, so we messed with that next.

Playing around with it, we found that the following diff fixes the problem (but, of course, also results in a bunch of unnecessary cache misses):

```
diff --git a/base/compiler/inferenceresult.jl b/base/compiler/inferenceresult.jl
index eb879f9211..99f5c83eae 100644
--- a/base/compiler/inferenceresult.jl
+++ b/base/compiler/inferenceresult.jl
@@ -140,7 +140,7 @@ function cache_lookup(linfo::MethodInstance, given_argtypes::Vector{Any}, cache:
                 a = maybe_widen_conditional(given_argtypes[i])
                 ca = cache_argtypes[i]
                 # verify that all Const argument types match between the call and cache
-                if is_argtype_mismatch(a, ca)
+                if isa(a, PartialTuple) || isa(b, PartialTuple) || is_argtype_mismatch(a, ca)
                     cache_match = false
                     break
                 end
```

From here we're a bit stuck...

vtjnash commented:
Actually, looking into this further, the cache lookup seems OK, but something is blocking us from doing optimization. Here's the code we spent the afternoon trying to look at this, but with the contextual information of inlining, so you can see which order it ran the optimizations:
```
     ┌ @ array.jl:769 within `setindex!'
     │┌ @ essentials.jl:275 within `convert'
16 ┄─││ %33  = (isa)(0x02, UInt64)::Bool
└────││        goto #27 if not %33
     ││ @ essentials.jl:275 within `convert' @ number.jl:7
     ││┌ @ boot.jl:716 within `Type'
     │││┌ @ boot.jl:631 within `toInt32'
     ││││┌ @ boot.jl:581 within `check_top_bit'
     │││││┌ @ boot.jl:571 within `is_top_bit_set'
17 ──││││││ %35  = (Core.lshr_int)(0x02, 63)::UInt64
│    ││││││ %36  = (Core.trunc_int)(Core.UInt8, %35)::UInt8
│    ││││││ %37  = (Core.eq_int)(%36, 0x01)::Bool
│    │││││┘
└────│││││        goto #19 if not %37
18 ──│││││        invoke Core.throw_inexacterror(:check_top_bit::Symbol, UInt64::Any, 0x02::UInt64)::Union{}
└────│││││        $(Expr(:unreachable))::Union{}
     │││││ @ boot.jl:582 within `check_top_bit'
19 ┄─│││││        goto #20
     ││││┘
     ││││┌ @ boot.jl:587 within `checked_trunc_sint'
20 ──│││││ %42  = (Core.trunc_int)(Int32, 0x02)::Int32
│    │││││ @ boot.jl:588 within `checked_trunc_sint'
│    │││││ %43  = (Core.sext_int)(UInt64, %42)::UInt64
│    │││││ @ boot.jl:589 within `checked_trunc_sint'
│    │││││ %44  = (Core.eq_int)(0x02, %43)::Bool
└────│││││        goto #22 if not %44
     │││││ @ boot.jl:590 within `checked_trunc_sint'
21 ──│││││        goto #23
     │││││ @ boot.jl:589 within `checked_trunc_sint'
22 ──│││││        invoke Core.throw_inexacterror(:trunc::Symbol, Int32::Any, 0x02::UInt64)::Union{}
└────│││││        $(Expr(:unreachable))::Union{}
     ││││┘
23 ┄─││││        goto #24
     │││┘
24 ──│││        goto #25
     ││┘
25 ──││        goto #26
     ││ @ essentials.jl:275 within `convert'
26 ──││        goto #30
27 ──││ %53  = (isa)(0x02, UInt8)::Bool
└────││        goto #29 if not %53
     ││ @ essentials.jl:275 within `convert' @ number.jl:7
     ││┌ @ boot.jl:716 within `Type'
     │││┌ @ boot.jl:628 within `toInt32'
28 ──││││ %55  = (Core.zext_int)(Core.Int32, 0x02)::Int32
│    ││┘┘
│    ││ @ essentials.jl:275 within `convert'
└────││        goto #30
29 ──││        (Core.throw)(ErrorException("fatal error in type inference (type bound)"))::Union{}
└────││        $(Expr(:unreachable))::Union{}
30 ┄─││ %59  = φ (#26 => %42, #28 => %55)::Int32
│    ││ %60  = (Core.tuple)(%59, %27)::Tuple{Int32,UInt64}
└────││        goto #31
     │┘
31 ──│        (Base.arrayset)(false, %7, %60, %14)::Array{Tuple{Int32,UInt64},1}
└────│        goto #32
     ┘
```

Keno commented:
> Actually, looking into this further, the cache lookup seems OK, but something is blocking us from doing optimization.

I don't understand what you mean by this.

jrevels commented:
So @vtjnash and I did another session here and made some more progress; it now builds and passes the tests added by this PR.

The InferenceResult cache code is also actually pretty nice now - the lookup and the construction code are starting to really converge. It might be a tad slower though, it deserves some benchmarking.

It still does not fix https://github.com/jrevels/Cassette.jl/issues/71,  but we are *really* close. We've narrowed that down to this MWE:

```
f(args...) = Core._apply(Core.tuple, args...)
g(x) = f((Bool, x))
```

Basically, inference should infer a `PartialTuple([Const(Bool), Int])` for the final result of `g(1)` (though it will of course actually return `Tuple{DataType,Int}` for e.g. printing purposes).

We have a patch which allows inference to accomplish this, and subsequently fixes https://github.com/jrevels/Cassette.jl/issues/71:

```
diff --git a/base/compiler/abstractinterpretation.jl b/base/compiler/abstractinterpretation.jl
index 3905763ca1..1656ae247c 100644
--- a/base/compiler/abstractinterpretation.jl
+++ b/base/compiler/abstractinterpretation.jl
@@ -1052,7 +1052,7 @@ function typeinf_local(frame::InferenceState)
             elseif hd === :return
                 pc´ = n + 1
                 rt = maybe_widen_conditional(abstract_eval(stmt.args[1], s[pc], frame))
-                if !isa(rt, Const) && !isa(rt, Type)
+                if !isa(rt, Const) && !isa(rt, Type) && !(isa(rt, PartialTuple) && frame.cached)
                     # only propagate information we know we can store
                     # and is valid inter-procedurally
                     rt = widenconst(rt)
```

However, this patch triggers a segfault (just run `make` to see the fireworks).

jrevels commented:
Assertion failures from local `make debug test`

```
      From worker 2:	Assertion failed: (arg.ispointer()), function emit_call_specfun_other, file /Users/jarrettrevels/data/repos/juliadev/src/codegen.cpp, line 2984.
      From worker 2:
      From worker 2:	signal (6): Abort trap: 6
      From worker 2:	in expression starting at /Users/jarrettrevels/data/repos/juliadev/test/abstractarray.jl:730
      From worker 2:	__pthread_kill at /usr/lib/system/libsystem_kernel.dylib (unknown line)
      From worker 2:	Allocations: 153547594 (Pool: 153527269; Big: 20325); GC: 344
Worker 2 terminated.
```

```
      From worker 3:	Assertion failed: (li->min_world <= world && li->max_world >= world), function jl_compile_linfo, file /Users/jarrettrevels/data/repos/juliadev/src/codegen.cpp, line 1091.
      From worker 3:
      From worker 3:	signal (6): Abort trap: 6
      From worker 3:	in expression starting at /Users/jarrettrevels/data/repos/juliadev/test/bitarray.jl:1578
      From worker 3:	__pthread_kill at /usr/lib/system/libsystem_kernel.dylib (unknown line)
      From worker 3:	Allocations: 416586215 (Pool: 416441975; Big: 144240); GC: 931
Worker 3 terminated.
```

Keno committed:
Turn some asserts into runtime traps instead

Unfortunately, we cannnot always rely on :invokes to have argument
values that match the declared ssa types (e.g. if the :invoke is
dynamically unreachable). Because of that, we cannot assert here,
but must instead emit a runtime trap.

jrevels commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

Keno commented:
I am unable to reproduce the bitarray CI failure.

jrevels commented:
With the most recent commit, this PR now fixes https://github.com/jrevels/Cassette.jl/issues/71 again

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/28955#issuecomment-427196846) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/2618eb6dfe6267a39e974cb1beb019df3af0b73c/4939aa5_vs_37ab710/report.md). cc @ararslan

jrevels commented:
Hmm...

Check out this example:

```
julia> x = rand(100, 100);

julia> f(x) = cumsum(x, dims=2)
f (generic function with 1 method)
```

On master:

```
julia> @time f(x);
  0.000033 seconds (11 allocations: 78.516 KiB)
```

Here:

```
julia> @time f(x);
  0.013461 seconds (149.31 k allocations: 3.267 MiB)
```

If we look at `@code_typed optimize=false` on both, they are exactly the same.

If we look at `@code_typed` on both, there are some extra `Expr(:unreachable)`s floating around on this branch. Diff: https://gist.github.com/jrevels/d7740244bfd07f70263f3ea9f6cefc46

If we look at `@code_llvm` on both, they are literally the same AFAICT (modulo some generated names, e.g. `julia__accumulate!_249602206` vs `julia__accumulate!_94174756` etc). Diff: https://gist.github.com/jrevels/0b37dd1f288ba0d4b1cd6888c38303de

Anybody see what's going on here? I don't...

Keno commented:
The actual code for the accumulate might be different?

jrevels commented:
D'oh, I need to get my first cup of coffee down before posting GitHub comments. Thanks.

jrevels commented:
Hmm. I see statements like this floating around in the IR on this branch (looking at `@code_typed Base._accumulate!(Base.add_sum, x, x, 2, nothing)`):

```
%93  = (isa)(1, Int64)::Bool
```

Looks like the recent widenconst additions to the isa_tfunc might've been too aggressive? I'll play around with it

Keno committed:
Don't eliminate PiNodes for mismatched types

E.g. if we had `PiNode(1, CartesianIndex)`, we would eliminate that
because `1` was a constant. However, leaving this in allows the compiler
to realize that this code is unreachable, as well as guarding codegen
against having to wrok through invalid IR.

Keno commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

Keno reviewed (approved):
FWIW, LGTM once CI and nanosoldier are ok.

Keno commented:
Ok, looks like this is basically passing. @jrevels Do you want to squash this down a bit (probably combine most of the PartialTuple stuff into one commit, but maybe leave all the things we found along the way - codegen changes, the PiNode thing, separate and prior in the commit list)?

jrevels committed:
Replace `InferenceResult` varargs hack with new inference lattice element for "partially constant" tuples

Previously, we hacked in an additional `InferenceResult` field to store varargs type information
in order to facilitate better constant propagation through varargs methods. There were many
other places, however, where constants moving in/out of tuples/varargs thwarted constant
propagation.

This commit removes the varargs hack, replacing it with a new inference lattice element
(`PartialTuple`) that represents tuples where some (but not all) of the elements are
constants. This allows us to follow through with constant propagation in more
situations involving tuple construction/destructuring, and also enabled a clean-up
of the `InferenceResult` caching code.

jrevels committed:
add tests for new PartialTuple capabilities

jrevels commented:
Squashed.

jrevels commented:
Argh. Not sure what went wrong but my rebase somehow messed something up. Time to consult the reflog...

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/28955#issuecomment-427455973) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/2a4c19566d59684f02d3e689ced66296ebac255c/78809be_vs_50c27fd/report.md). cc @ararslan

Keno commented:
Looks good to me. @jrevels wanna do the honors?

jrevels commented:
Thanks again to @vtjnash @Keno and @JeffBezanson for your help here!
